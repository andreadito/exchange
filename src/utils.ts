export enum Currency {
    EUR = 'EUR',
    GBP = 'GBP',
    USD = 'USD',
}

export interface WalletData {
    balance: number;
    currency: Currency;
}

export interface Wallet {
    id: string,
    data: WalletData
}

export interface Rate {
    symbol: string;
    value: string;
}

export const MIN_ALLOWED = 0.5;

export function convertValue(value: number, rate: number): number {
    return value * rate;
}

export function limitDecimal(value: string) {
    return (value.indexOf(".") >= 0) ? (value.substr(0, value.indexOf(".")) + value.substr(value.indexOf("."), 3)) : value
}
