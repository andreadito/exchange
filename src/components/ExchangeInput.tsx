import React from "react";
import {Box, Select, Text, TextInput} from "grommet";

import {Controller} from "react-hook-form";
import {ErrorMessage} from "@hookform/error-message";
import {Control} from "react-hook-form/dist/types/form";
import {limitDecimal, MIN_ALLOWED, Wallet} from "../utils";

export interface ExchangeInputProps {
    name: string;
    label: string;
    isExceeding?: boolean;
    wallets: Wallet[];
    control: Control;
    errors: any;
    rate?: number;
    readonly?: boolean
}

export default function ExchangeInput({name, label, wallets, control, errors, isExceeding, rate = 1, readonly}: ExchangeInputProps) {
    const onInput = (e: { currentTarget: { value: any; }; }) => {
        const value = e.currentTarget.value;
        e.currentTarget.value = limitDecimal(value);
    };
    return (
        <Box>
            <Box gap={'small'} fill={'horizontal'} direction={'row'} alignContent={'between'}>
                <Controller control={control} render={({onChange, name, value}) => {
                    return (
                        <Box width={'small'}>
                            <Select
                                plain={readonly}
                                placeholder={label}
                                name={name}
                                options={wallets}
                                labelKey={(wallet: Wallet) => `${wallet.data.currency}`}
                                size={'medium'}
                                value={value}
                                onChange={({value}) => onChange(value)}
                            />
                            {value && value.data && (<Text
                                color={isExceeding ? 'status-critical' : 'status-ok'}
                                style={{marginTop: 10}} size={'xsmall'}>
                                {isExceeding ? 'Is exceeding your balance' : `Balance: ${value.data.balance.toFixed(2)}`}
                            </Text>)
                            }
                        </Box>
                    )
                }} name={`${name}_wallet`}/>
                <Controller rules={{
                    min: {
                        value: !readonly ? MIN_ALLOWED : 0,
                        message: `Should be greater that ${MIN_ALLOWED}`
                    },
                    required: !readonly,
                }} name={`${name}_value`} control={control} render={({onChange, name, value}) => {

                    const fixedValues = readonly ? value.toFixed(2) : value;

                    return (
                        <TextInput
                            plain={readonly}
                            name={name}
                            type={'number'}
                            inputMode={"decimal"}
                            step={0.01}
                            size={'medium'}
                            value={fixedValues}
                            onChange={onChange}
                            style={{textAlign: 'right'}}
                            placeholder={"0"}
                            readOnly={readonly}
                            onInput={onInput}
                        />
                    )
                }}/>
            </Box>
            {!readonly && (
                <ErrorMessage
                    errors={errors}
                    name={`${name}_value`}
                    render={({message}) => (
                        <Text color={'status-error'} alignSelf={'end'} size={'xsmall'}>{message}</Text>)}
                />
            )}
        </Box>
    )
}
