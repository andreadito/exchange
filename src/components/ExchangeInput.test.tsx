import ExchangeInput from "./ExchangeInput";
import React from "react";
import {Currency, Wallet} from "../utils";
import {useForm} from "react-hook-form";
import renderer from 'react-test-renderer';
import { renderHook } from "@testing-library/react-hooks";
import 'mutationobserver-shim';

describe("Exchange Input", () => {
    const wallets: Wallet[] = [
        {
            "id": "AyMzv3EkKFzFG4H8HfEz",
            "data": {
                "currency": Currency.GBP,
                "balance": 1090
            }
        },
        {
            "id": "SPNFH5Q5pDLBbYOPrOd6",
            "data": {
                "currency": Currency.EUR,
                "balance": 0
            }
        },
        {
            "id": "amAienpx5MVa5EJkyJEc",
            "data": {
                "balance": 10,
                "currency": Currency.USD
            }
        }
    ];

    it("should render correctly", () => {
        const { result } = renderHook(() => useForm({
            defaultValues:{
                from_value: 0,
                from_wallet: wallets[0],
                to_value: 0,
                to_wallet: wallets[1],
            },
            mode: "onChange"
        }))
        const tree = renderer.create(<ExchangeInput
            name={'from'}
            label={'From'}
            wallets={wallets}
            control={result.current.control}
            errors={[]}/>).toJSON();
        expect(tree).toMatchSnapshot();
    })
    it("should render correctly if exceeding", () => {
        const { result } = renderHook(() => useForm({
            defaultValues:{
                from_value: 0,
                from_wallet: wallets[0],
                to_value: 0,
                to_wallet: wallets[1],
            },
            mode: "onChange"
        }))
        const tree = renderer.create(<ExchangeInput
            name={'from'}
            label={'From'}
            wallets={wallets}
            control={result.current.control}
            errors={[]}
            isExceeding={true}/>).toJSON();
        expect(tree).toMatchSnapshot();
    })
    it("should render correctly if a validation error occour", () => {
        const { result } = renderHook(() => useForm({
            defaultValues:{
                from_value: 0,
                from_wallet: wallets[0],
                to_value: 0,
                to_wallet: wallets[1],
            },
            mode: "onChange"
        }))
        const tree = renderer.create(<ExchangeInput
            name={'from'}
            label={'From'}
            wallets={wallets}
            control={result.current.control}
            errors={{
                "from_value": {
                    "type": "min",
                    "message": "Should be greater that 0.5",
                    "ref": {
                        "name": "from_value"
                    }
                },
                "to_value": {
                    "type": "min",
                    "message": "Should be greater that 0.5",
                    "ref": {
                        "name": "to_value"
                    }
                }
            }}
            isExceeding={true}/>).toJSON();
        expect(tree).toMatchSnapshot();
    })
});
