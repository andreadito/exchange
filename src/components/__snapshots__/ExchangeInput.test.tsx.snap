// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange Input should render correctly 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 gAsOLC"
>
  <div
    className="StyledBox-sc-13pk1d4-0 fkEaZX"
  >
    <div
      className="StyledBox-sc-13pk1d4-0 kJuWxj"
    >
      <button
        aria-label="Open Drop"
        className="StyledButton-sc-323bzc-0 kUDdtj Select__StyledSelectDropButton-sc-17idtfo-1 bsheiM"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        type="button"
      >
        <div
          className="StyledBox-sc-13pk1d4-0 lhHpuy"
        >
          <div
            className="StyledBox-sc-13pk1d4-0 gErjJY"
          >
            <div
              className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 flNVco"
            >
              <input
                autoComplete="off"
                className="StyledTextInput-sc-1x30a0s-0 eymfXE Select__SelectTextInput-sc-17idtfo-0 jwiVPE"
                name="from_wallet"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="From"
                readOnly={true}
                size="medium"
                tabIndex="-1"
                type="text"
                value="GBP"
              />
            </div>
          </div>
          <div
            className="StyledBox-sc-13pk1d4-0 dderd"
            style={
              Object {
                "minWidth": "auto",
              }
            }
          >
            <svg
              aria-label="FormDown"
              className="StyledIcon-ofa7kd-0 gvGxfc"
              viewBox="0 0 24 24"
            >
              <polyline
                fill="none"
                points="18 9 12 15 6 9"
                stroke="#000"
                strokeWidth="2"
              />
            </svg>
          </div>
        </div>
      </button>
      <span
        className="StyledText-sc-1sadyjn-0 CVDRa"
        style={
          Object {
            "marginTop": 10,
          }
        }
      >
        Balance: 1090
      </span>
    </div>
    <div
      className="StyledBox__StyledBoxGap-sc-13pk1d4-1 bZqeeV"
    />
    <div
      className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 flNVco"
    >
      <input
        autoComplete="off"
        className="StyledTextInput-sc-1x30a0s-0 ksHnCo"
        inputMode="decimal"
        name="from_value"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onInput={[Function]}
        onKeyDown={[Function]}
        placeholder="0"
        size="medium"
        step={0.01}
        style={
          Object {
            "textAlign": "right",
          }
        }
        type="number"
        value={0}
      />
    </div>
  </div>
</div>
`;

exports[`Exchange Input should render correctly if a validation error occour 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 gAsOLC"
>
  <div
    className="StyledBox-sc-13pk1d4-0 fkEaZX"
  >
    <div
      className="StyledBox-sc-13pk1d4-0 kJuWxj"
    >
      <button
        aria-label="Open Drop"
        className="StyledButton-sc-323bzc-0 kUDdtj Select__StyledSelectDropButton-sc-17idtfo-1 bsheiM"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        type="button"
      >
        <div
          className="StyledBox-sc-13pk1d4-0 lhHpuy"
        >
          <div
            className="StyledBox-sc-13pk1d4-0 gErjJY"
          >
            <div
              className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 flNVco"
            >
              <input
                autoComplete="off"
                className="StyledTextInput-sc-1x30a0s-0 eymfXE Select__SelectTextInput-sc-17idtfo-0 jwiVPE"
                name="from_wallet"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="From"
                readOnly={true}
                size="medium"
                tabIndex="-1"
                type="text"
                value="GBP"
              />
            </div>
          </div>
          <div
            className="StyledBox-sc-13pk1d4-0 dderd"
            style={
              Object {
                "minWidth": "auto",
              }
            }
          >
            <svg
              aria-label="FormDown"
              className="StyledIcon-ofa7kd-0 gvGxfc"
              viewBox="0 0 24 24"
            >
              <polyline
                fill="none"
                points="18 9 12 15 6 9"
                stroke="#000"
                strokeWidth="2"
              />
            </svg>
          </div>
        </div>
      </button>
      <span
        className="StyledText-sc-1sadyjn-0 epPaeH"
        style={
          Object {
            "marginTop": 10,
          }
        }
      >
        Is exceeding your balance
      </span>
    </div>
    <div
      className="StyledBox__StyledBoxGap-sc-13pk1d4-1 bZqeeV"
    />
    <div
      className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 flNVco"
    >
      <input
        autoComplete="off"
        className="StyledTextInput-sc-1x30a0s-0 ksHnCo"
        inputMode="decimal"
        name="from_value"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onInput={[Function]}
        onKeyDown={[Function]}
        placeholder="0"
        size="medium"
        step={0.01}
        style={
          Object {
            "textAlign": "right",
          }
        }
        type="number"
        value={0}
      />
    </div>
  </div>
  <span
    className="StyledText-sc-1sadyjn-0 iFPVEX"
  >
    Should be greater that 0.5
  </span>
</div>
`;

exports[`Exchange Input should render correctly if exceeding 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 gAsOLC"
>
  <div
    className="StyledBox-sc-13pk1d4-0 fkEaZX"
  >
    <div
      className="StyledBox-sc-13pk1d4-0 kJuWxj"
    >
      <button
        aria-label="Open Drop"
        className="StyledButton-sc-323bzc-0 kUDdtj Select__StyledSelectDropButton-sc-17idtfo-1 bsheiM"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        type="button"
      >
        <div
          className="StyledBox-sc-13pk1d4-0 lhHpuy"
        >
          <div
            className="StyledBox-sc-13pk1d4-0 gErjJY"
          >
            <div
              className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 flNVco"
            >
              <input
                autoComplete="off"
                className="StyledTextInput-sc-1x30a0s-0 eymfXE Select__SelectTextInput-sc-17idtfo-0 jwiVPE"
                name="from_wallet"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="From"
                readOnly={true}
                size="medium"
                tabIndex="-1"
                type="text"
                value="GBP"
              />
            </div>
          </div>
          <div
            className="StyledBox-sc-13pk1d4-0 dderd"
            style={
              Object {
                "minWidth": "auto",
              }
            }
          >
            <svg
              aria-label="FormDown"
              className="StyledIcon-ofa7kd-0 gvGxfc"
              viewBox="0 0 24 24"
            >
              <polyline
                fill="none"
                points="18 9 12 15 6 9"
                stroke="#000"
                strokeWidth="2"
              />
            </svg>
          </div>
        </div>
      </button>
      <span
        className="StyledText-sc-1sadyjn-0 epPaeH"
        style={
          Object {
            "marginTop": 10,
          }
        }
      >
        Is exceeding your balance
      </span>
    </div>
    <div
      className="StyledBox__StyledBoxGap-sc-13pk1d4-1 bZqeeV"
    />
    <div
      className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 flNVco"
    >
      <input
        autoComplete="off"
        className="StyledTextInput-sc-1x30a0s-0 ksHnCo"
        inputMode="decimal"
        name="from_value"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onInput={[Function]}
        onKeyDown={[Function]}
        placeholder="0"
        size="medium"
        step={0.01}
        style={
          Object {
            "textAlign": "right",
          }
        }
        type="number"
        value={0}
      />
    </div>
  </div>
</div>
`;
